### PR TITLE
mentions: Make @-mentions tappable to navigate to user profiles

### LINKED ISSUE #1867: Tap an @-mention to go to user's profile page
In web, if you click on an @-mention of a user, it shows that user's "user card", just as if you'd clicked on their name or avatar as the sender of a message.

In mobile, if you tap on a sender's name or avatar, we take you to the user's profile page. But currently we don't do anything on tapping @-mentions. We should do the same thing on tapping an @-mention as we do on tapping a sender name or avatar.

### References
* User feedback in chat: [#mobile > changed alias not updated in messages @ 💬](https://chat.zulip.org/#narrow/channel/48-mobile/topic/changed.20alias.20not.20updated.20in.20messages/near/2264980)

### PR DESCRIPTION
**Pull request by AshutoshKhadse23**
Demo:

https://github.com/user-attachments/assets/fd7a960f-0f73-4ba2-935b-4a9b22cb9dbc

Closes #1867

### CODE REVIEW

**Comment by gnprice** [comment]
Thanks. Because you marked this PR as a draft, we haven't looked at it for review. There's now a second PR open for this issue, #2083.

Each of the PRs has some elements that are better than the other. So I think a productive next step, before maintainer review, would be for each of you and @chungwwei to review the other's PR. Then you can try to produce a co-authored version that has the better aspects of both.

File: test/widgets/content_test.dart [hunk 1]
```diff
@@ -820,6 +821,53 @@ void main() {
...
+      // `tester.pumpWidget` in prepareContent introduces an initial route;
+      // remove it so consumers only have newly pushed routes.
+      assert(pushedRoutes.length == 1);
+      pushedRoutes.removeLast();
+      return pushedRoutes;
+    }
+
+    testWidgets('tapping on user mention navigates to profile page', (tester) async {
+      final pushedRoutes = await prepareForTapTest(tester,
+        '<p><span class="user-mention" data-user-id="2187">@Greg Price</span></p>');
```
Comment (chungwwei) [inline test/widgets/content_test.dart]: Let's reuse the examples like `ContentExample.userMentionPlain` in `model/content_test.dart` and put these tests in a group

File: test/model/content_test.dart [hunk 2]
```diff
@@ -111,105 +111,105 @@ class ContentExample {
     "@**Greg Price**",
     expectedText: '@Greg Price',
     '<p><span class="user-mention" data-user-id="2187">@Greg Price</span></p>',
-    const UserMentionNode(nodes: [TextNode('@Greg Price')]));
+    const UserMentionNode(userId: 2187, nodes: [TextNode('@Greg Price')]));
```
Comment (chungwwei) [inline test/model/content_test.dart]: Put `userId` after `nodes` since that's order the class is defined

File: lib/model/content.dart [hunk 3]
```diff
@@ -1120,11 +1129,21 @@ class _ZulipInlineContentParser {
...
 
+    // Extract user ID for individual user mentions.
+    // - Individual user mentions have data-user-id="<numeric_id>"
+    // - Wildcard mentions have data-user-id="*"
+    // - Group mentions have data-user-group-id instead
+    // - Topic mentions don't have a user ID attribute
+    final userIdStr = element.attributes['data-user-id'];
+    final int? userId = (userIdStr != null && userIdStr != '*')
+      ? int.tryParse(userIdStr)
+      : null;
```
Comment (chungwwei) [inline lib/model/content.dart]: `final int? userId = int.tryParse(element.attributes['data-user-id'] ?? '');`

checking is not necessary, `tryParse` will return `null` on a not integer in a string literal.

File: lib/model/content.dart [hunk 4]
```diff
@@ -1120,11 +1129,21 @@ class _ZulipInlineContentParser {
       return null;
     }
 
+    // Extract user ID for individual user mentions.
+    // - Individual user mentions have data-user-id="<numeric_id>"
+    // - Wildcard mentions have data-user-id="*"
+    // - Group mentions have data-user-group-id instead
+    // - Topic mentions don't have a user ID attribute
```
Comment (chungwwei) [inline lib/model/content.dart]: I don't think these comments are necessary

**Comment by chungwwei** [comment]
Apologies, I pushed my PR without being aware of this PR. I’ve closed the other PR, but feel free to review it if you’d like.

Here are a few things I noticed that could be fixed or improved. Each commit should pass the existing tests. It's probably in the contributor guide.
- Adding the `userId` breaks the existing tests, those tests should go with the commit that adds the `userId` field. 
- The tests that are for the navigation should go with the commit that adds the functionality.
- I would remove the long body text in the commit messages body, the code changes are small enough to tell the story.

**Comment by AshutoshKhadse23** [comment]
@chungwwei, I have considered your comment and made changes.
Thank you for your feedback

File: test/model/content_test.dart [hunk 5]
```diff
@@ -111,105 +111,105 @@ class ContentExample {
...
-    const UserMentionNode(nodes: [TextNode('@test-empty')]));
+    const UserMentionNode(userId: null, nodes: [TextNode('@test-empty')]));
 
   static final groupMentionSilent = ContentExample.inline(
     'silent group @-mention',
     "@_*test-empty*",
     expectedText: 'test-empty',
     '<p><span class="user-group-mention silent" data-user-group-id="186">test-empty</span></p>',
-    const UserMentionNode(nodes: [TextNode('test-empty')]));
+    const UserMentionNode(userId: null, nodes: [TextNode('test-empty')]));
```
Comment (chungwwei) [inline test/model/content_test.dart]: bump https://github.com/zulip/zulip-flutter/pull/2075#discussion_r2729822992.  Put `userId` after `nodes` for all of them.

File: test/widgets/content_test.dart [hunk 6]
```diff
@@ -820,6 +821,55 @@ void main() {
     // TODO(#647):
     //  testFontWeight('non-silent self-user mention in bold context',
     //    expectedWght: 800, // [etc.]
+
+    Future<List<Route<dynamic>>> prepareForTapTest(WidgetTester tester, String html) async {
+      final pushedRoutes = <Route<dynamic>>[];
```
Comment (chungwwei) [inline test/widgets/content_test.dart]: Move it inside the group, it's only written for the navigation tests right?

File: test/widgets/content_test.dart [hunk 7]
```diff
@@ -820,6 +821,55 @@ void main() {
...
+        final pushedRoutes = await prepareForTapTest(tester,
+          ContentExample.groupMentionPlain.html);
+
+        await tester.tap(find.text('@test-empty'));
+        check(pushedRoutes).isEmpty();
+      });
+
+      testWidgets('tapping on wildcard mention does not navigate', (tester) async {
+        final pushedRoutes = await prepareForTapTest(tester,
+          '<p><span class="user-mention channel-wildcard-mention" data-user-id="*">@all</span></p>');
```
Comment (chungwwei) [inline test/widgets/content_test.dart]: bump https://github.com/zulip/zulip-flutter/pull/2075#discussion_r2729822545. 

The commit message lists out these cases.
```
- Tapping user mention navigates to profile page
- Tapping group mention does not navigate
- Tapping wildcard mention does not navigate
- Tapping topic mention does not navigate
```

Since there are already many `ContentExample` defined, we can reuse those for the tests.

**Comment by chungwwei** [comment]
Thanks for the revision.

bump the bullet points from https://github.com/zulip/zulip-flutter/pull/2075#issuecomment-3802567501 and read https://github.com/zulip/zulip-flutter/pull/2087#issuecomment-3786970872. PR #2087 could be a good reference for how to structure commits and tests for this issue.

nit: the commit bodies might be a bit long for these three commits — the summary lines already seem sufficient.

File: lib/widgets/content.dart [hunk 8]
```diff
@@ -1258,17 +1275,18 @@ class UserMention extends StatelessWidget {
       padding: const EdgeInsets.symmetric(horizontal: 0.2 * kBaseFontSize),
       child: InlineContent(
         // If an @-mention is inside a link, let the @-mention override it.
-        recognizer: null,  // TODO(#1867) make @-mentions tappable, for info on user
+        // For individual user mentions, navigate to their profile page on tap.
```
Comment (chungwwei) [inline lib/widgets/content.dart]: This comment doesn’t seem helpful and could be removed.

**Comment by chungwwei** [comment]
Hey @AshutoshKhadse23 — still up for putting the PR together? Now that #2087 is merged, this should be much smoother from here

**Comment by chungwwei** [comment]
I see that you've pushed changes, thank you. I have noted a few areas where we can make improvements, see below.

557e780 can be dropped from this pr, since the other pr already handled it, we don't need to modify any comments.
286fbf2 looks good
dec416e needs work. I have revised it to the following, could you please see below code block and see if it's a improvement to your code. We should add more test cases(look at `testContentSmoke` in `group('UserMention'`) and handle the edge case where the user is unknown.

dec416e needs to be squashed into 286fbf2. Tests relevant to the feature should stay in the same commit the feature is on. If you need help on rebasing, I can help you on https://chat.zulip.org/#narrow/channel/44-git-help
```dart 
    group('UserMention tap test', () {
      void tapMentionTest({
        required String description,
        required String html,
        int? expectedUserId,
      }) {
        testWidgets(description, (tester) async {
          final pushedRoutes = <Route<dynamic>>[];
          final testNavObserver = TestNavigatorObserver()
            ..onPushed = (route, prevRoute) => pushedRoutes.add(route);

          await prepareContent(tester,
            plainContent(html), navObservers: [testNavObserver],
            wrapWithPerAccountStoreWidget: true);

          assert(pushedRoutes.length == 1);
          pushedRoutes.removeLast();
          await tester.tap(find.byType(UserMention));
          if (expectedUserId != null) {
            check(pushedRoutes).single
              .isA<WidgetRoute>().page
              .isA<ProfilePage>().userId
              .equals(expectedUserId);
          } else {
            check(pushedRoutes).isEmpty();
          }
        });
      }

      tapMentionTest(description: 'tapping on user mention navigates to profile page',
        html: ContentExample.userMentionPlain.html, expectedUserId: 2187);
      tapMentionTest(description: 'tapping on silent user mention navigates to profile page',
        html: ContentExample.userMentionSilent.html, expectedUserId: 2187);
      tapMentionTest(description: 'tapping groupMentionPlain does nothing',
        html: ContentExample.groupMentionPlain.html, expectedUserId: null);
      
      // ...
      // Can you please add the rest of the tests and handle the edge case Unkown User
    });
```

File: test/widgets/content_test.dart [hunk 9]
```diff
@@ -1091,26 +1092,93 @@ void main() {
...
+          html: '<p><span class=\"user-mention channel-wildcard-mention\" data-user-id=\"*\">@all</span></p>');
         check(find.text('@all')).findsOne();
       });
 
       testWidgets('handles silent mentions correctly', (tester) async {
         await prepare(
           tester: tester,
-          html: '<p><span class="user-mention silent" data-user-id="123">Old Name</span></p>',
+          html: '<p><span class=\"user-mention silent\" data-user-id=\"123\">Old Name</span></p>',
           users: [eg.selfUser, eg.user(userId: 123, fullName: 'New Name')]);
```
Comment (chungwwei) [inline test/widgets/content_test.dart]: Could you please remove the escape backslashes from all of them?

message from running the analyzer in CI
```
   info • Unnecessary escape in string literal • test/widgets/content_test.dart:1109:33 • unnecessary_string_escapes
   info • Unnecessary escape in string literal • test/widgets/content_test.dart:1109:54 • unnecessary_string_escapes
   info • Unnecessary escape in string literal • test/widgets/content_test.dart:1109:70 • unnecessary_string_escapes
```

**Comment by AshutoshKhadse23** [comment]
@chungwwei 
I need to make some changes after the other pull request got merged could you check that if ok then I will make it ready as review

File: test/widgets/content_test.dart [hunk 10]
```diff
@@ -1111,6 +1112,77 @@ void main() {
...
+        await prepareContent(tester,
+          plainContent('<p><span class="user-mention" data-user-id="999">@Unknown User</span></p>'),
+          navObservers: [testNavObserver],
+          wrapWithPerAccountStoreWidget: true);
+
+        assert(pushedRoutes.length == 1);
+        pushedRoutes.removeLast();
+        await tester.tap(find.byType(UserMention));
+        check(pushedRoutes).isEmpty();
+      });
```
Comment (chungwwei) [inline test/widgets/content_test.dart]: If a user is not found, `ProfilePage` builds `_ProfileErrorPage`.

https://github.com/zulip/zulip-flutter/blob/ec45f1eb53e3f6406795174b352a0a624651b91a/lib/widgets/profile.dart#L50-L56

We can just use the function defined above. Function doesn’t explicitly check that `_ProfileErrorPage` is built, but I think navigating to `ProfilePage` should be good enough.
```suggestion
      tapMentionTest(description: 'tapping on an unknown user navigates to ProfilePage, which builds a _ProfileErrorPage',
        html: '<p><span class="user-mention" data-user-id="999">@Unknown User</span></p>', expectedUserId: 999);
```

File: lib/widgets/content.dart [hunk 11]
```diff
@@ -1192,15 +1193,64 @@ class UserMention extends StatelessWidget {
...
+
+class _UserMentionState extends State<UserMention> {
+  GestureRecognizer? _recognizer;
+
+  void _prepareRecognizer() {
+    final userId = widget.node.userId;
+    if (userId != null) {
+      final store = PerAccountStoreWidget.of(context);
+      final user = store.getUser(userId);
+      if (user != null) {
```
Comment (chungwwei) [inline lib/widgets/content.dart]: I'd revert the function to what you had before. `_ProfileErrorPage` is built when a user is not found.

**Comment by chungwwei** [comment]
@AshutoshKhadse23 Thanks for the update. Aside from the user not found case, I think it looks good.

hmmm, `tapMentionTest` probably isn’t the best name when I suggested it earlier, I had used `userMentionNavigationTest` in my old PR.

File: lib/widgets/content.dart [hunk 12]
```diff
@@ -1192,15 +1193,64 @@ class UserMention extends StatelessWidget {
...
+    _recognizer?.dispose();
+    _recognizer = null;
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _disposeRecognizer();
+    _prepareRecognizer();
+  }
```
Comment (chungwwei) [inline lib/widgets/content.dart]: I missed this change. Why was this added, and `initState()` is removed?

**Comment by AshutoshKhadse23** [comment]
@chungwwei Now, what do you think? According to me, we made it ready to merge?

**Comment by chungwwei** [comment]
~~From my side, this looks ready for maintainer review.~~

A related PR has just been merged, so this branch will need to be rebased. The conflicts don’t seem too bad to resolve.

**Comment by AshutoshKhadse23** [comment]
> ~From my side, this looks ready for maintainer review.~
> 
> A related PR has just been merged, so this branch will need to be rebased. The conflicts don’t seem too bad to resolve.

@chungwwei, I have completed rebasing. I think it is now ready for maintainer review

File: lib/widgets/content.dart [hunk 13]
```diff
@@ -1192,23 +1193,66 @@ class Mention extends StatelessWidget {
...
+  GestureRecognizer? _recognizer;
+
+  void _prepareRecognizer() {
+    if (widget.node case UserMentionNode(:final userId?)) {
+      _recognizer = TapGestureRecognizer()
+        ..onTap = () {
+          Navigator.push(context,
+            ProfilePage.buildRoute(context: context, userId: userId));
+        };
+    }
```
Comment (chungwwei) [inline lib/widgets/content.dart]: Since `MentionNode` is a sealed class, it might make more sense to use a `switch` statement here. similar to how it’s handled in `build()`.

Comment (AshutoshKhadse23) [reply lib/widgets/content.dart]: yes did it

**Comment by chungwwei** [comment]
Thanks @AshutoshKhadse23. LGTM.

**Comment by AshutoshKhadse23** [comment]
@chrisbobbe, We have refined this PR, and as the other PR is in integration review, I think I should convert this to ready to review.

File: lib/widgets/content.dart [hunk 14]
```diff
@@ -1221,14 +1268,14 @@ class Mention extends StatelessWidget {
       padding: const EdgeInsets.symmetric(horizontal: 0.2 * kBaseFontSize),
       child: InlineContent(
         // If an @-mention is inside a link, let the @-mention override it.
-        recognizer: null,  // TODO(#1867) make @-mentions tappable, for info on user
+        recognizer: _recognizer,
```
Comment (chrisbobbe) [inline lib/widgets/content.dart]: Instead of handling taps with this `InlineContent.recognizer` mechanism, what if we just added a `GestureDetector` wrapper around the `Container` in this build method? Something like this:

```diff
@@ -1213,7 +1214,7 @@ class Mention extends StatelessWidget {
       case UserMentionNode(userId: null):
     }
 
-    return Container(
+    Widget result = Container(
       decoration: BoxDecoration(
         // TODO(#646) different for wildcard mentions
         color: contentTheme.colorDirectMentionBackground,
@@ -1231,6 +1232,17 @@ class Mention extends StatelessWidget {
         style: ambientTextStyle,
 
         nodes: nodes));
+
+    if (node case UserMentionNode(:final userId?)) {
+      result = GestureDetector(
+        onTap: () {
+          Navigator.of(context).push(
+            ProfilePage.buildRoute(context: context, userId: userId));
+        },
+        child: result);
+    }
+
+    return result;
   }
 
 // This is a more literal translation of Zulip web's CSS.
```

That seems much simpler, and would also make the @-mention's horizontal padding respond to taps, not just the text itself.

File: test/widgets/content_test.dart [hunk 15]
```diff
@@ -1150,6 +1151,63 @@ void main() {
...
+            ]));
+
+          assert(pushedRoutes.length == 1);
+          pushedRoutes.removeLast();
+          await tester.tap(find.byType(Mention));
+          if (expectedUserId != null) {
+            check(pushedRoutes).single
+              .isA<WidgetRoute>().page
+              .isA<ProfilePage>().userId
+              .equals(expectedUserId);
```
Comment (chrisbobbe) [inline test/widgets/content_test.dart]: nit: put "page is a profile page" on its own line; "user ID is the expected one" on its own line, e.g. like this:

```suggestion
            check(pushedRoutes).single.isA<WidgetRoute>()
              .page.isA<ProfilePage>()
              .userId.equals(expectedUserId);
```

File: test/widgets/content_test.dart [hunk 16]
```diff
@@ -1150,6 +1151,63 @@ void main() {
...
+          }
+        });
+      }
+
+      userMentionNavigationTest(description: 'tapping on user mention navigates to profile page',
+        html: ContentExample.userMentionPlain.html, expectedUserId: 2187);
+      userMentionNavigationTest(description: 'tapping on silent user mention navigates to profile page',
+        html: ContentExample.userMentionSilent.html, expectedUserId: 2187);
+      userMentionNavigationTest(description: 'tapping on an unknown user navigates to ProfilePage, which builds a _ProfileErrorPage',
+        html: '<p><span class="user-mention" data-user-id="999">@Unknown User</span></p>', expectedUserId: 999);
```
Comment (chrisbobbe) [inline test/widgets/content_test.dart]: The description makes it sound like "builds a _ProfileErrorPage" is part of what the test will check, but the test doesn't check that. How about de-emphasizing that by being less specific and using parentheses, like:

> tapping on an unknown user navigates to ProfilePage (which handles this case)

File: test/widgets/content_test.dart [hunk 17]
```diff
@@ -1150,6 +1151,63 @@ void main() {
...
+      userMentionNavigationTest(description: 'tapping groupMentionPlain does nothing',
+        html: ContentExample.groupMentionPlain.html, expectedUserId: null);
```
Comment (chrisbobbe) [inline test/widgets/content_test.dart]: nit: write the description according to the interface being targeted in the test, which is the human/user interface:

```suggestion
      userMentionNavigationTest(description: 'tapping plain group @-mention does nothing',
        html: ContentExample.groupMentionPlain.html, expectedUserId: null);
```

(for all of these)

File: test/widgets/content_test.dart [hunk 18]
```diff
@@ -1150,6 +1151,63 @@ void main() {
...
+      userMentionNavigationTest(description: 'tapping channelWildcardMentionSilent does nothing',
+        html: ContentExample.channelWildcardMentionSilent.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping legacyChannelWildcardMentionPlain does nothing',
+        html: ContentExample.legacyChannelWildcardMentionPlain.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping legacyChannelWildcardMentionSilent does nothing',
+        html: ContentExample.legacyChannelWildcardMentionSilent.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping topicMentionPlain does nothing',
+        html: ContentExample.topicMentionPlain.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping topicMentionSilent does nothing',
+        html: ContentExample.topicMentionSilent.html, expectedUserId: null);
```
Comment (chrisbobbe) [inline test/widgets/content_test.dart]: Some of these tests aren't about user mentions, but "UserMention" is in the `group`'s description. Can we harmonize that? Should the `group` have a different name or should some of these tests go in a different `group`?

**Review by chrisbobbe — COMMENTED** [review]
Thanks! Comments below.

File: lib/widgets/content.dart (hunk 14 above)
Comment (chungwwei) [reply lib/widgets/content.dart]: Ahh! `Mention` is a child of `WidgetSpan`. That means we don’t need a `GestureRecognizer` like we would for a `TextSpan`.
```dart
case MentionNode():
  return WidgetSpan(alignment: PlaceholderAlignment.middle,
    child: Mention(ambientTextStyle: widget.style, node: node));
```

**Comment by AshutoshKhadse23** [comment]
@chrisbobbe, I have made the necessary changes and is ready to review

File: test/widgets/content_test.dart [hunk 19]
```diff
@@ -1226,6 +1227,62 @@ void main() {
...
+              .page.isA<ProfilePage>()
+              .userId.equals(expectedUserId);
+          } else {
+            check(pushedRoutes).isEmpty();
+          }
+        });
+      }
+
+      userMentionNavigationTest(description: 'tapping on user mention navigates to profile page',
+        html: ContentExample.userMentionPlain.html, expectedUserId: 2187);
```
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: `description` and `html` can be positional parameters; when reading a call site, it's clear enough what they mean just from seeing the argument values

File: test/widgets/content_test.dart [hunk 20]
```diff
@@ -1226,6 +1227,62 @@ void main() {
...
+      userMentionNavigationTest(description: 'tapping on silent user mention navigates to profile page',
+        html: ContentExample.userMentionSilent.html, expectedUserId: 2187);
```
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: a helper that makes a test case (i.e. calls `test` or `testWidgets`) should signal that to the reader by having a name starting with `test`; so e.g. `testUserMentionNavigation`

File: test/widgets/content_test.dart [hunk 21]
```diff
@@ -1226,6 +1227,62 @@ void main() {
...
+      userMentionNavigationTest(description: 'tapping on an unknown user navigates to ProfilePage (which handles this case)',
+        html: '<p><span class="user-mention" data-user-id="999">@Unknown User</span></p>', expectedUserId: 999);
+      userMentionNavigationTest(description: 'tapping plain group @-mention does nothing',
+        html: ContentExample.groupMentionPlain.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping silent group @-mention does nothing',
+        html: ContentExample.groupMentionSilent.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping plain channel wildcard @-mention does nothing',
+        html: ContentExample.channelWildcardMentionPlain.html, expectedUserId: null);
```
Comment (gnprice) [inline test/widgets/content_test.dart]: But also these aren't all user mentions, right? So a better name might be just `testMentionNavigation`.

File: test/widgets/content_test.dart [hunk 22]
```diff
@@ -1226,6 +1227,62 @@ void main() {
...
+      userMentionNavigationTest(description: 'tapping silent channel wildcard @-mention does nothing',
+        html: ContentExample.channelWildcardMentionSilent.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping plain legacy channel wildcard @-mention does nothing',
+        html: ContentExample.legacyChannelWildcardMentionPlain.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping silent legacy channel wildcard @-mention does nothing',
+        html: ContentExample.legacyChannelWildcardMentionSilent.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping plain topic @-mention does nothing',
+        html: ContentExample.topicMentionPlain.html, expectedUserId: null);
+      userMentionNavigationTest(description: 'tapping silent topic @-mention does nothing',
+        html: ContentExample.topicMentionSilent.html, expectedUserId: null);
```
Comment (gnprice) [inline test/widgets/content_test.dart]: We can make this a lot simpler to read, and less work to maintain in the future, by cutting out the majority of these test cases.

The key purpose of a test is to catch when a future change would introduce a bug. For most of these, it's hard to imagine what sort of bug would be caught by any one of these tests that wouldn't already have been caught by the others.

Specifically, I think it's enough to have:
* the three tests for actual user mentions
* one for a plain group mention
* one for a wildcard mention: either a plain channel mention or a plain topic mention

The type of bug that the latter two might conceivably catch is if the code tried to somehow treat one of those mentions as if it were a user mention, and then threw an error or did something else odd. We're checking that the code instead does nothing with them. For that it's enough to check the most boring, common version of each type of mention: if we correctly do nothing with those, it's unlikely we'll accidentally be doing something with the more arcane versions.

File: test/widgets/content_test.dart [hunk 23]
```diff
@@ -1226,6 +1227,62 @@ void main() {
...
+          }
+        });
+      }
+
+      userMentionNavigationTest(description: 'tapping on user mention navigates to profile page',
+        html: ContentExample.userMentionPlain.html, expectedUserId: 2187);
+      userMentionNavigationTest(description: 'tapping on silent user mention navigates to profile page',
+        html: ContentExample.userMentionSilent.html, expectedUserId: 2187);
+      userMentionNavigationTest(description: 'tapping on an unknown user navigates to ProfilePage (which handles this case)',
+        html: '<p><span class="user-mention" data-user-id="999">@Unknown User</span></p>', expectedUserId: 999);
```
Comment (gnprice) [inline test/widgets/content_test.dart]: Hmm, what makes one of these an unknown user and the other a known user?

… Ah, it's that the helper code above treats 2187 as a special value by adding a user with that ID.

That's a less clear way to write the test than we prefer to do: it means that there's an important aspect of what these tests are testing (namely, whether the user is known or not, which is the point of this last test compared to the two before it) which the reader can't see from looking at the individual test cases. They can only see it by reading through the details of the helper code.

Better would be for the tests that need a particular user to exist to be explicit about adding that user. That can be done with an optional argument to the test helper. It looks like only two of these test cases would need it.

**Review by gnprice — COMMENTED** [review]
Thanks @AshutoshKhadse23 and @chungwwei, and @chrisbobbe for the previous review!

Generally this looks good. Some comments below on the tests.

**Comment by AshutoshKhadse23** [comment]
@gnprice, thank you for the comments. I have implemented them and am waiting for any other comments.

File: test/widgets/content_test.dart [hunk 24]
```diff
@@ -1226,6 +1227,61 @@ void main() {
...
+              .page.isA<ProfilePage>()
+              .userId.equals(expectedUserId);
+          } else {
+            check(pushedRoutes).isEmpty();
+          }
+        });
+      }
+
+      testMentionNavigation(
+        'tapping on user mention navigates to profile page',
```
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: this formatting was good for compactness (and matches the usual way of calling `test` or `testWidgets`):
```suggestion
      testMentionNavigation('tapping on user mention navigates to profile page',
```

File: test/widgets/content_test.dart [hunk 25]
```diff
@@ -1226,6 +1227,61 @@ void main() {
...
+        ContentExample.userMentionPlain.html,
+        expectedUserId: 2187,
+        addUser: eg.user(userId: 2187, fullName: 'Greg Price'));
```
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: put the setup first, then the expected result after that — that way it matches the logical order of how these get used

File: test/widgets/content_test.dart [hunk 26]
```diff
@@ -1226,6 +1227,61 @@ void main() {
...
+      testMentionNavigation(
+        'tapping on silent user mention navigates to profile page',
+        ContentExample.userMentionSilent.html,
+        expectedUserId: 2187,
+        addUser: eg.user(userId: 2187, fullName: 'Greg Price'));
```
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: the user's name doesn't matter for this test, right? We can leave the name out, to keep the test focused on only the details that are relevant for the reader to understand what it's doing.

(The user object needs a name, but the `eg.user` helper will make up some name; that's its job, filling in details that are needed for making the data realistic but are boring for a given test.)

File: test/widgets/content_test.dart [hunk 27]
```diff
@@ -1226,6 +1227,61 @@ void main() {
...
+      testMentionNavigation(
+        'tapping on an unknown user navigates to ProfilePage (which handles this case)',
+        '<p><span class="user-mention" data-user-id="999">@Unknown User</span></p>',
+        expectedUserId: 999);
+      testMentionNavigation(
+        'tapping plain group @-mention does nothing',
+        ContentExample.groupMentionPlain.html);
```
Comment (gnprice) [inline test/widgets/content_test.dart]: The "does nothing" part is key to what this test is checking for. So it should make that explicit; the `expectedUserId: null` in the previous revision was useful.

File: test/widgets/content_test.dart [hunk 28]
```diff
@@ -1226,6 +1227,61 @@ void main() {
...
+          }
+        });
+      }
+
+      testMentionNavigation(
+        'tapping on user mention navigates to profile page',
+        ContentExample.userMentionPlain.html,
+        expectedUserId: 2187,
+        addUser: eg.user(userId: 2187, fullName: 'Greg Price'));
+      testMentionNavigation(
```
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: put blank lines between each of these calls, as they're each multiple lines long; when they were all 2 lines it was marginal, but now that they're longer they should be separated.

In particular, and especially combined with putting the descriptions back on the first line, that will make it easier to scan the list and see all the descriptions.

File: test/widgets/content_test.dart (hunk 25 above)
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: the name "add user", being a verb phrase, sounds like it's going to be the name of a callback function; since it's just a piece of plain data, a noun (or noun phrase) would be better, like just "user"

**Review by gnprice — COMMENTED** [review]
Thanks for the revision! Comments below.

Note there's also an error flagged by CI.

**Comment by AshutoshKhadse23** [comment]
@gnprice, Thanks for the comments. I have solved the error flagged by CI and implemented the changes as per the comments.

File: test/widgets/content_test.dart [hunk 29]
```diff
@@ -1226,6 +1227,62 @@ void main() {
...
       });
     });
+
+    group('@-mention tap tests', () {
+      void testMentionNavigation(
+        String description,
+        String html, {
+        User? user,
+        int? expectedUserId,
+      }) {
```
Comment (gnprice) [inline test/widgets/content_test.dart]: nit: following up on https://github.com/zulip/zulip-flutter/pull/2075#discussion_r3029980882 , this parameter should be required, since all callers should be explicit about it

**Review by gnprice — COMMENTED** [review]
Thanks everyone for all your work on this! All looks good modulo one nit below. I'll fix that and merge.
